pkg_resources/__init__.py

```diff
@@ -2492,7 +2492,7 @@ def resolve(self):
             raise ImportError(str(exc)) from exc
 
     def require(self, env=None, installer=None):
-        if self.extras and not self.dist:
+        if not self.dist:
```

Comment:
The issue here is the code `self.dist.requires` below, which isn't type-safe id `self.dist` is potentially `None`.
The change of behaviour is that now a `UnknownExtra` will be raised if you have no distribution. Which I guess can be weird if you have no extras. The message is still appropriate though. A change like this would keep the `AttributeError` from before, be type-safe, explicit about value validation, and have an appropriate error message. (from the user's pov, only the message would change)
```suggestion
        if not self.dist:
            error_cls = UnknownExtra if self.extras else AttributeError
            raise error_cls("Can't require() without a distribution", self)
```

Reply:
That sounds reasonable. Thanks @Avasam.